core/trino-main/src/main/java/io/trino/metadata/MetadataListing.java

```diff
@@ -248,61 +248,76 @@ private static Map<SchemaTableName, List<ColumnMetadata>> doListTableColumns(Ses
+                    QualifiedObjectName actualTableName;
+                    TableHandle targetTableHandle;
+                    try {
+                        // For redirected tables, column listing requires special handling, because the column metadata is unavailable
+                        // at the source table, and needs to be fetched from the target table.
+                        RedirectionAwareTableHandle redirection = metadata.getRedirectionAwareTableHandle(session, originalTableName);
+
+                        // The target table name should be non-empty. If it is empty, it means that there is an
+                        // inconsistency in the connector's implementation of ConnectorMetadata#streamTableColumns and
+                        // ConnectorMetadata#redirectTable.
+                        if (redirection.redirectedTableName().isEmpty()) {
+                            return;
+                        }
+                        actualTableName = redirection.redirectedTableName().get();
+                        targetTableHandle = redirection.tableHandle().orElseThrow();
                     }
-                    actualTableName = redirection.redirectedTableName().get();
-                    targetTableHandle = redirection.tableHandle().orElseThrow();
-                }
-                catch (TrinoException e) {
-                    // Ignore redirection errors
-                    if (e.getErrorCode().equals(TABLE_REDIRECTION_ERROR.toErrorCode())) {
-                        return;
+                    catch (TrinoException e) {
+                        // Ignore redirection errors
+                        if (e.getErrorCode().equals(TABLE_REDIRECTION_ERROR.toErrorCode())) {
+                            return;
+                        }
+                        throw e;
                     }
-                    throw e;
-                }
-
-                columns = metadata.getTableMetadata(session, targetTableHandle).getColumns();
-            }
 
-            Set<String> allowedColumns = accessControl.filterColumns(
-                    session.toSecurityContext(),
-                    // Use redirected table name for applying column filters, since the source does not know the column metadata
-                    actualTableName.asCatalogSchemaTableName(),
-                    columns.stream()
-                            .map(ColumnMetadata::getName)
-                            .collect(toImmutableSet()));
-            result.put(
-                    table,
-                    columns.stream()
-                            .filter(column -> allowedColumns.contains(column.getName()))
-                            .collect(toImmutableList()));
-        });
+                    List<ColumnMetadata> columns = metadata.getTableMetadata(session, targetTableHandle).getColumns();
+
+                    Set<String> allowedColumns = accessControl.filterColumns(
```

Comment:
Can all redirected tables be collected and batched in a single `accessControl.filterColumns` call?